R/assert.R

```diff
@@ -178,20 +163,129 @@ vec_is <- function(x, ptype = NULL, size = NULL) {
   TRUE
 }
 
-#' Is object a vector?
-#' @noRd
+#' Vector checks
 #'
 #' @description
 #'
-#' Returns `TRUE` if:
+#' - `vec_is_vector()` tests if `x` is considered a vector in the vctrs sense.
+#'   See _Vectors and scalars_ below for the exact details.
+#'
+#' - `vec_check_vector()` uses `vec_is_vector()` and throws a standardized and
+#'   informative error if it returns `FALSE`.
+#'
+#' - `vec_check_size()` tests if `x` has the same size as `size`, and throws
+#'   an informative error if it doesn't.
+#'
+#' @inheritParams rlang::args_dots_empty
+#' @inheritParams rlang::args_error_context
+#'
+#' @param x An object.
+#'
+#' @param size The size to check for.
+#'
+#' @returns
+#' - `vec_is_vector()` returns a single `TRUE` or `FALSE`.
+#'
+#' - `vec_check_vector()` returns `NULL` invisibly, or errors.
+#'
+#' - `vec_check_size()` returns `NULL` invisibly, or errors.
+#'
+#' @section Vectors and scalars:
+#'
+#' Informally, a vector is a collection that makes sense to use as column in a
+#' data frame. The following rules define whether or not `x` is considered a
+#' vector.
+#'
+#' If no [vec_proxy()] method has been registered, `x` is a vector if:
+#'
+#' - The [base type][typeof] of the object is atomic: `"logical"`, `"integer"`,
+#'   `"double"`, `"complex"`, `"character"`, or `"raw"`.
 #'
-#' * `x` is an atomic, whether it has a class or not.
-#' * `x` is a bare list without class.
-#' * `x` implements [vec_proxy()].
+#' - `x` is a list, as defined by [vec_is_list()].
 #'
-#' S3 lists are thus treated as scalars unless they implement a proxy.
+#' - `x` is a [data.frame].
+#'
+#' If a `vec_proxy()` method has been registered, `x` is a vector if:
+#'
+#' - The proxy satisfies one of the above conditions.
+#'
+#' - The base type of the proxy is `"list"`, regardless of its class. S3 lists
+#'   are thus treated as scalars unless they implement a `vec_proxy()` method.
```

Comment:
oh yeah.

To me requiring bare lists is a bit redundant because if we see a proxy and get a list, we should be confident that it's a vector. It'd be weird to treat a type that has a classed list proxy as a scalar. Also checking for a classed list will make the implementation slower not faster, since checking for a base type requires fewer pointer hops than checking for the presence of an attribute. Or am I missing something?

But I'm not feeling strongly about this. I mainly worry about the overhead of unclassing. If you're confident that requiring a full copy of the address vector for S3 lists like list-of will not be a performance problem, we could reconsider requiring bare lists.